Adds Raycaster with tracking for Dynamic Meshes

renezurbruegg opened the pull request:
# Description

This PR introduces `MultiMeshRayCaster` and `MultiMeshRayCasterCamera`, an extension of the default `RayCaster` with the following enhancements:

1. **Raycasting against multiple target types** : Supports primitive shapes (spheres, cubes, …) as well as arbitrary meshes.
2. **Dynamic mesh tracking** : Keeps track of specified meshes, enabling raycasting against moving parts (e.g., robot links, articulated bodies, or dynamic obstacles).
3. **Memory-efficient caching** : Avoids redundant memory usage by caching and reusing duplicate meshes.

This is joint work with @pascal-roth and @Mayankm96.

The default `RayCaster` was limited to static environments and required manual handling of moving meshes, which restricted its use for robotics scenarios where robots or obstacles move dynamically.  

`MultiMeshRayCaster` addresses these limitations by and now supports  **raycasting against robot parts and other moving entities**.

---
## Usage

For a quick demo, run:

```bash
python scripts/demos/sensors/multi_mesh_raycaster.py --num_envs 16 --asset_type <allegro_hand|anymal_d|multi>
```

<img width="2882" height="804" alt="demo image" src="https://github.com/user-attachments/assets/a019e9d4-e991-4ca2-a94e-8cdba790f26d" />

### Drop-in replacement

Example change to migrate from `RayCasterCfg` to `MultiMeshRayCasterCfg`:

```diff
- ray_caster_cfg = RayCasterCfg(
+ ray_caster_cfg = MultiMeshRayCasterCfg(
      prim_path="{ENV_REGEX_NS}/Robot",
-     mesh_prim_paths=[
-        "/World/ground"
-     ],
+     mesh_prim_paths=[
+         "/World/Ground",
+         MultiMeshRayCasterCfg.RaycastTargetCfg(target_prim_expr="{ENV_REGEX_NS}/Robot/LF_.*/visuals"),
+         MultiMeshRayCasterCfg.RaycastTargetCfg(target_prim_expr="{ENV_REGEX_NS}/Robot/RF_.*/visuals"),
+         MultiMeshRayCasterCfg.RaycastTargetCfg(target_prim_expr="{ENV_REGEX_NS}/Robot/LH_.*/visuals"),
+         MultiMeshRayCasterCfg.RaycastTargetCfg(target_prim_expr="{ENV_REGEX_NS}/Robot/RH_.*/visuals"),
+         MultiMeshRayCasterCfg.RaycastTargetCfg(target_prim_expr="{ENV_REGEX_NS}/Robot/base/visuals"),
+     ],
      pattern_cfg=patterns.GridPatternCfg(resolution=resolution, size=(5.0, 5.0)),
+     track_mesh_transforms=True,
 )
```

---

## Benchmarking & Validation

To benchmark the new raycaster, run:

```bash
python scripts/benchmarks/benchmark_ray_caster.py
```

Then plot the results with:

```bash
python scripts/benchmarks/plot_raycast_results.py
```

This will generate outputs under:
`outputs/benchmarks/raycast_benchmark...`

### Example plots

<table style="border-collapse:collapse; width:100%;">
  <!-- Row 1: big image across both columns -->
  <tr>
    <td colspan="2" style="text-align:center; padding-bottom:8px;">
      <img width="1000" height="500" alt="big image" src="https://github.com/user-attachments/assets/ff69253c-0329-4ab6-a944-7fcaac233923" />
    </td>
  </tr>

  <!-- Row 2: two images side by side -->

  <tr>
    <td style="text-align:center; padding-right:8px;">
      <img width="500" height="500" alt="left image" src="https://github.com/user-attachments/assets/5375ed6c-f419-448d-ba25-759c1b16bcdd" />
    </td>
    <td style="text-align:center;">
      <img width="500" height="500" alt="right image" src="https://github.com/user-attachments/assets/bc36a0a6-aedd-4cdb-9909-9a05b1f2be0e" />
    </td>
  </tr>

  <!-- Row 3: last image centered across both columns -->

  <tr>
    <td colspan="2" style="text-align:center; padding-top:8px;">
      <img width="500" height="500" alt="bottom image" src="https://github.com/user-attachments/assets/67b75944-f64e-4c0b-b51f-aa20da3cf9b1" />
    </td>
  </tr>
</table>

---

## Type of Change

* [x] New feature (non-breaking change which adds functionality)
* [ ] This change requires a documentation update



## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

kellyguo11 commented on scripts/benchmarks/plot_raycast_results.py:
```diff
@@ -0,0 +1,186 @@
+
+def compare_num_assets():
+    """Plots Average steps per ms vs Num assets and Avg memory vs num assets"""
+    df = pd.read_csv("/workspace/isaaclab/outputs/benchmarks/ray_caster_benchmark_num_assets.csv")
```
same comment about path

Mayankm96 commented on source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster_cfg.py:
```diff
@@ -0,0 +1,62 @@
+        Not tracking the mesh transformations is recommended when the meshes are static to increase performance.
+    """
+
+    merge_prim_meshes: bool = True
+    """Whether to merge the meshes under each entry of :attr:`mesh_prim_paths`."""
```
Should this property live inside `RaycastTargetCfg`?

Mayankm96 commented on scripts/demos/sensors/muli_mesh_raycaster.py:
```diff
@@ -0,0 +1,290 @@
+        offset=MultiMeshRayCasterCfg.OffsetCfg(pos=(0, -0.1, 0.3)),
+        mesh_prim_paths=[
+            "/World/Ground",
+            MultiMeshRayCasterCfg.RaycastTargetCfg(target_prim_expr="{ENV_REGEX_NS}/Robot/LF_.*/visuals"),
```
Where is {ENV_REGEX_NS} getting resolved?

Mayankm96 commented on source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster_cfg.py:
```diff
@@ -0,0 +1,66 @@
+        """The regex to specify the target prim to ray cast against."""
+
+        is_global: bool = False
+        """Whether the target prim is a global object or exists for each environment instance. Defaults to False."""
```
As I was writing more explanation, I was wondering if the expression resolves to only 1 prim, is it safe to assume that it is a global target? Otherwise, let's say, user defines this attribute to `False` but passes "/World/ground". Would in this case, this only be used to the first env?

```suggestion
        """Whether the target prim is a global object or exists for each environment instance. Defaults to False.
        
        If True, the target prim is used for ray-casting in all environments.
        Otherwise, it is only used per environment.
        """
```

renezurbruegg replied:
Very good point! I think right now, it just fails tbh.

Mayankm96 commented on source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster_cfg.py:
```diff
@@ -0,0 +1,66 @@
+        """Whether the target prim is assumed to be the same mesh across all environments. Defaults to False.
+
+        If True, only the first mesh is read and then reused for all environments, rather than re-parsed.
+        This provides a startup performance boost when there are many environments that all use the same asset. Note,
+        if reference_meshes is False, this flag has no effect.
```
```suggestion
        This provides a startup performance boost when there are many environments that all use the same asset.
        
        .. note::
            If :attr:`MultiMeshRayCasterCfg.reference_meshes` is False, this flag has no effect.
```

Mayankm96 commented on source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster_cfg.py:
```diff
@@ -0,0 +1,66 @@
+        track_mesh_transforms: bool = True
+        """Whether the mesh transformations should be tracked. Defaults to True.
+
+        Note:
```
```suggestion
        .. note::
```

Mayankm96 commented on source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster_cfg.py:
```diff
@@ -0,0 +1,66 @@
+    class_type: type = MultiMeshRayCaster
+
+    mesh_prim_paths: list[str | RaycastTargetCfg] = MISSING
+    """The list of mesh primitive paths to ray cast against. If an entry is a string, it is internally converted to a :class:`RaycastTargetCfg` with global set to true and track_mesh_transforms disabled to ensure backwards compatibility with the default raycaster."""
```
```suggestion
    """The list of mesh primitive paths to ray cast against.
    
    If an entry is a string, it is internally converted to :class:`RaycastTargetCfg` with :attr:`~RaycastTargetCfg.is_global` enabled and attr:`~RaycastTargetCfg.track_mesh_transforms` disabled. These settings ensure backwards compatibility with the default raycaster.
    """
```

Mayankm96 commented on source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster_camera.py:
```diff
@@ -0,0 +1,162 @@
+    def __str__(self) -> str:
+        """Returns: A string containing information about the instance."""
+        return (
+            f"Ray-Caster-Camera @ '{self.cfg.prim_path}': \n"
```
```suggestion
            f"Multi-Mesh Ray-Caster-Camera @ '{self.cfg.prim_path}': \n"
```

Mayankm96 commented on source/isaaclab/isaaclab/sensors/utils.py:
```diff
@@ -0,0 +1,47 @@
+from isaaclab.utils.math import convert_quat
+
+
+def compute_world_poses(
```
Instead of opening a new `utils.py`, can we just move these to `sim/utils.py` and rename it to `obtain_world_pose_from_view` or something?

Mayankm96 commented on docs/source/api/lab/isaaclab.utils.rst:
```diff
@@ -20,6 +20,7 @@
       timer
       types
       warp
+      mesh
```
Nice. I think we will need to consolidate the terrain mesh utils also in this to just have one place with bunch of stuff. Right now things have become spreaded out...

Mayankm96 commented on source/isaaclab/isaaclab/sim/utils.py:
```diff
@@ -198,7 +201,7 @@ def wrapper(prim_path: str | Sdf.Path, *args, **kwargs):
             # this is based on the physics behavior that nested schemas are not allowed
             if not success:
-                all_prims += child_prim.GetChildren()
+                all_prims += child_prim.GetFilteredChildren(Usd.TraverseInstanceProxies())
```
Hm apply_nested won't work on instanced proxies so parsing them here is unnecssary. users have to make the asset uninstanced before they can do this function.

renezurbruegg replied:
Merging in your PR instead of these changes

Mayankm96 commented on source/isaaclab/isaaclab/sim/utils.py:
```diff
@@ -570,7 +573,72 @@ def make_uninstanceable(prim_path: str | Sdf.Path, stage: Usd.Stage | None = Non
+    return world_position, world_orientation
+
+
+def resolve_relative_pose(
```
is it cleaner to just keep one of the two  and have optional arg for relative to?

renezurbruegg replied:
Merging in your PR instead of these changes

Mayankm96 commented on source/isaaclab/isaaclab/sim/utils.py:
```diff
@@ -570,7 +573,72 @@ def make_uninstanceable(prim_path: str | Sdf.Path, stage: Usd.Stage | None = Non
+    if relative_to is None:
+        return torch.tensor(prim_pos, device=device), torch.tensor(prim_quat, device=device)
+
+    relative_to_pos, relative_to_quat = resolve_world_pose(relative_to)
```
Device isn't resolved correctly.

Mayankm96 replied:
Instead of torch operation, we can also simply use pxr GfMat4 math

renezurbruegg replied:
Merging in your PR instead of these changes

Mayankm96 commented on source/isaaclab/isaaclab/sim/utils.py:
```diff
@@ -570,7 +573,72 @@ def make_uninstanceable(prim_path: str | Sdf.Path, stage: Usd.Stage | None = Non
+        A tuple containing the world position (as a 3D vector) and the world orientation (as a quaternion, format wxyz).
+    """
+
+    xform = UsdGeom.Xformable(prim)
+    world_transform = xform.ComputeLocalToWorldTransform(Usd.TimeCode.Default())
+    world_position = [*world_transform.ExtractTranslation()]
+    world_orientation = [
+        world_transform.ExtractRotation().GetQuaternion().GetNormalized().real,
+        *world_transform.ExtractRotation().GetQuaternion().GetNormalized().imaginary,
+    ]
```
In my tests, this returned wrong values. Please check: https://github.com/isaac-sim/IsaacLab/pull/3371

kellyguo11 commented:
`test_multi_mesh_ray_caster_camera.py` tests are failing in CI. could you also run `./isaaclab.sh -f` when the changes are ready pls?

Mayankm96 commented on source/isaaclab/isaaclab/utils/mesh.py:
```diff
@@ -0,0 +1,132 @@
+    # Load faces and convert to triangle if needed. (Default is quads)
+    num_vertex_per_face = np.asarray(mesh.GetFaceVertexCountsAttr().Get())
+    indices = np.asarray(mesh.GetFaceVertexIndicesAttr().Get())
+    return points, convert_faces_to_triangles(indices, num_vertex_per_face)
```
Should return a trimesh mesh instance to match the function name?

kellyguo11 commented:
hmm looks like there's still a failure in CI :(
```
| /workspace/isaaclab/source/isaaclab/test/sensors/test_multi_mesh_ray_caster_camera.py  | FAILED |    30.52 |  12/21  |
```

ooctipus commented:
as an example

some robot has prim_path of `{ENV_REGEX_NS}/Robot` where the ArticulationRootAPI is applied
other robot may have ArticulationRootAPI applied at prim_path `{ENV_REGEX_NS}/Robot/root_joint`

renezurbruegg commented:
@pascal-roth if you find some time, maybe you can update the MultiMeshRayCasterCamera to also use to `_update_rays_impl` workflow, which updates the ray directions and sensor origin

ooctipus commented:
I also see that I have a same triangulate face utility https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/dexsuite/mdp/utils.py

```
def _triangulate_faces(prim) -> np.ndarray:
    mesh = UsdGeom.Mesh(prim)
    counts = mesh.GetFaceVertexCountsAttr().Get()
    indices = mesh.GetFaceVertexIndicesAttr().Get()
    faces = []
    it = iter(indices)
    for cnt in counts:
        poly = [next(it) for _ in range(cnt)]
        for k in range(1, cnt - 1):
            faces.append([poly[0], poly[k], poly[k + 1]])
    return np.asarray(faces, dtype=np.int64)
```

your:
```
def convert_faces_to_triangles(faces: np.ndarray, point_counts: np.ndarray) -> np.ndarray:
    # check if the mesh is already triangulated
    if (point_counts == 3).all():
        return faces.reshape(-1, 3)  # already triangulated
    all_faces = []

    vertex_counter = 0
    # Iterates over all triangles of the mesh.
    # could be very slow for large meshes
    for num_points in point_counts:
        if num_points == 3:
            # triangle
            all_faces.append(faces[vertex_counter : vertex_counter + 3])
        elif num_points == 4:
            # quads. Subdivide into two triangles
            f = faces[vertex_counter : vertex_counter + 4]
            first_triangle = f[:3]
            second_triangle = np.array([f[0], f[2], f[3]])
            all_faces.append(first_triangle)
            all_faces.append(second_triangle)
        else:
            raise RuntimeError(f"Invalid number of points per face: {num_points}")

        vertex_counter += num_points
    return np.asarray(all_faces)
```

I wonder maybe you want the code still to work beyond num_points > 4?

ooctipus commented on source/isaaclab/isaaclab/sim/utils.py:
```diff
@@ -568,7 +573,93 @@ def make_uninstanceable(prim_path: str | Sdf.Path, stage: Usd.Stage | None = Non
+    # extract position and orientation
+    prim_pos = [*prim_tf.ExtractTranslation()]
+    prim_quat = [prim_tf.ExtractRotationQuat().real, *prim_tf.ExtractRotationQuat().imaginary]
+    return tuple(prim_pos), tuple(prim_quat)
```
be sure to use xform_cache it give better performance : ))

Mayankm96 commented:
This is definitely useful feedback @pavelacamposp 

@renezurbruegg I suggest though that we keep this MR to the current limitation and make a separate one with the hashing implementation. Just to not have this MR hanging around for too long :)

renezurbruegg commented:
I believe I addressed all relevant comments! 

FYI: While cleaning up some warp kernels, I found a bug in the warp operators for atomic_min https://github.com/NVIDIA/warp/issues/1058. I believe this bug is quite severe and will hopefully be addressed soon. For now I added a fixme to the warp kernel sections.

Mayankm96 commented:
@kellyguo11 I made an MR now to address the comments remaining here. I think Pascal and Rene are OoO now.
